docs: add DD-478 Bitcoin > ICP for Bitcoin page (#3057)
dfinity/portal: docs/developer-docs/multi-chain/bitcoin/using-btc/btc-comparison.mdx

---
keywords: [intermediate, bitcoin, btc, icp]
---
import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

# Understanding ICP for Bitcoin developers

<MarkdownChipRow labels={["Intermediate", "Bitcoin"]} />

## Overview
This guide is designed to help Bitcoin developers understand the core concepts of ICP for using ICP as a Layer 2 solution for Bitcoin and as a separate blockchain.

### Smart contracts

> Bitcoin’s scripting language, Script, is fairly simple, which makes complex smart contracts rather difficult to deploy directly on the base layer. 

Reviewer: ```suggestion
Bitcoin’s scripting language, Script, makes complex smart contracts rather difficult to deploy directly on the Bitcoin mainnet due to the language's limited capabilities.
```

> Bitcoin developers who are looking to build more complex smart contracts will need to build on a Bitcoin smart contract layer such as ICP, each offering its unique set of developer tools. 

Reviewer: ```suggestion
Bitcoin developers who are looking to build complex smart contracts will need to use Bitcoin smart contract layer (referred to as a Layer-2 or L2), such as ICP. Each L2 solution offers a unique set of developer tools. 
```

Author: How about "Bitcoin developers who are looking to build complex smart contracts will need to use a Bitcoin smart contract layer"?

> Canister smart contracts on ICP have special properties that allow developers to develop scalable services. Canister smart contracts can be written in any language that complies to WASM including [Rust](/docs/current/developer-docs/backend/rust/), [Typescript](/docs/current/developer-docs/backend/Typescript), [Motoko](/docs/current/motoko/main/getting-started/motoko-introduction), and [Python](/docs/current/developer-docs/backend/python/). This further allows for greater complexity in smart contract development. 

Reviewer: ```suggestion
Canister smart contracts on ICP have special properties that allow developers to develop scalable services. Canisters can be written in any language that complies to Wasm including [Rust](/docs/current/developer-docs/backend/rust/), [Typescript](/docs/current/developer-docs/backend/Typescript), [Motoko](/docs/current/motoko/main/getting-started/motoko-introduction), and [Python](/docs/current/developer-docs/backend/python/). This allows for greater complexity in smart contract development compared to Bitcoin's Script language.
```

> To learn more about full canister smart contract functionality, refer to the [Smart Contracts documentation](/docs/current/developer-docs/smart-contracts/overview/introduction).

Reviewer: ```suggestion
To learn more about the full capabilities of canister smart contracts](/docs/current/developer-docs/smart-contracts/overview/introduction).
```

> Specifically, through protocol-level integration of ICP with the [Bitcoin network](docs/current/developer-docs/multi-chain/bitcoin/overview), ICP smart contracts can interact with the Bitcoin network directly. Canister smart contracts can run [queries against the full Bitcoin UTXO](/docs/current/developer-docs/multi-chain/bitcoin/using-btc/read-state) set including finding the balance of any Bitcoin address. They can [hold](/docs/current/developer-docs/multi-chain/bitcoin/using-btc/generate-addresses), [sign](/docs/current/developer-docs/multi-chain/bitcoin/using-btc/sign-transactions), and [submit transactions](/docs/current/developer-docs/multi-chain/bitcoin/using-btc/submit-transactions) to the Bitcoin network. 

Reviewer: ```suggestion

## Direct integration
Through the protocol-level integration of ICP and the [Bitcoin network](docs/current/developer-docs/multi-chain/bitcoin/overview), canisters can interact with the Bitcoin network directly for workflows such as:

- Run [queries against the full Bitcoin UTXO](/docs/current/developer-docs/multi-chain/bitcoin/using-btc/read-state) set, including finding the balance of any Bitcoin address. 

- [Hold BTC](/docs/current/developer-docs/multi-chain/bitcoin/using-btc/generate-addresses).

- [Sign](/docs/current/developer-docs/multi-chain/bitcoin/using-btc/sign-transactions) and [submit transactions](/docs/current/developer-docs/multi-chain/bitcoin/using-btc/submit-transactions) directly to the Bitcoin network. 
```

Reviewer: Move the 'Direct integration' section above 'Smart contracts'

> To learn more about the Bitcoin integration on ICP, refer to the [Bitcoin integration documentation](docs/current/developer-docs/multi-chain/bitcoin/overview).
> 
> Additionally, the upcoming [Threshold Schnorr integration](https://forum.dfinity.org/t/threshold-schnorr-facilitating-brc-20-trading-solana-integration-certificate-signing-and-more/28993) will allow developers to interact with Runes and Ordinals directly in canister smart contracts. A [Schnorr canister example](https://github.com/domwoe/schnorr_canister/) is available as a preview of this integration.
> 

Reviewer: ```suggestion

The upcoming [Threshold Schnorr integration](https://forum.dfinity.org/t/threshold-schnorr-facilitating-brc-20-trading-solana-integration-certificate-signing-and-more/28993) will allow developers to interact with Runes and Ordinals directly in canisters. A [Schnorr canister example](https://github.com/domwoe/schnorr_canister/) is available as a preview of this integration.

To learn more about the Bitcoin integration on ICP, refer to the [Bitcoin integration documentation](docs/current/developer-docs/multi-chain/bitcoin/overview).
```

> ICP uses different terminology for concepts similar to those in Bitcoin. This table summarizes the key terminology differences between Bitcoin and ICP:

Reviewer: ```suggestion

## Terminology 

ICP uses different terminology for concepts similar to those in Bitcoin. This table summarizes the key terminology differences between Bitcoin and ICP:
```

<table>
<thead>
<tr>
  <th> Bitcoin term </th>
  <th> ICP term </th>
</tr>
</thead>

<tbody>
<tr>
  <td> Smart contract </td>
  <td> Canister </td>
</tr>

<tr>
  <td> Gas </td>
  <td> Cycles </td>
</tr>
  <td> Account </td>
  <td> Principal </td>
</tr>

</tbody>
</table>

In addition, ICP has implemented similar token standards as Bitcoin:

<table>
<thead>
<tr>
  <th> Bitcoin standard </th>
  <th> ICP standard </th>
</tr>
</thead>

<tbody>
<tr>
  <td> BRC-20 </td>
  <td> [ICRC-1/ICRC-2](../../../../references/icrc1-standard) </td>
</tr>

<tr>
  <td> BRC-721 </td>
  <td> [ICRC-7](https://github.com/dfinity/ICRC/blob/icrc7-wg-draft/ICRCs/ICRC-7/ICRC-7.md) </td>
</tr>
</tbody>
</table>

### Consensus

> Bitcoin uses Proof of Work as its consensus mechanism.

Reviewer: ```suggestion
Bitcoin uses proof of work as its consensus mechanism. 
```

Reviewer: Maybe add some additional context about the downsides of PoW- high finality times, expensive transactions, etc

> The IC consensus protocol provides cryptographically guaranteed finality. The approach on ICP avoids the drawbacks of probabilistic finality used in Bitcoin-like protocols, which include the weak notion and increased time to finality. The consensus mechanism of ICP ensures faster and more secure transaction finality.

Reviewer: Instead of using the same wording as in the ETH doc, maybe try to highlight how ICP's consensus allows for significantly improved dev experience as a BTC L2, since transaction times are much faster and cheaper than on the BTC mainnet.

Author: Thank you for the feedback. This is a good idea! Can you please check the Consensus section now and see if you have any other feedback?

Reviewer: LGTM!

> The IC consensus protocol provides cryptographically guaranteed finality. The approach on ICP avoids the drawbacks of probabilistic finality used in Bitcoin-like protocols, which include the weak notion and increased time to finality. The consensus mechanism of ICP ensures faster and more secure transaction finality.

Reviewer: ```suggestion
The IC consensus protocol provides cryptographically guaranteed finality. The approach on ICP avoids the drawbacks of probabilistic finality used in Bitcoin-like protocols, which include the weak notion and increased time to finality. The consensus mechanism of ICP ensures faster and more secure transaction finality.

## Resources

[Add resources]
```

To learn more about the Bitcoin integration on ICP, refer to the [Bitcoin integration documentation](docs/current/developer-docs/multi-chain/bitcoin/overview).

### Consensus
Bitcoin uses proof of work as its consensus mechanism. Using proof of work, the Bitcoin blockchain waits for ten minutes before validating transactions one block at a time, posing challenges to security and scalability.

Instead of imposing time constraints, ICP relies of cryptographic signatures to finalize a block. A block maker selects available messages, merging them into a block. The consensus protocol maintains a tree of notarized blocks. Each replica creates a distinct finalization share for every notarized block, serving as a definitive indicator that consensus has been achieved. This entire process could take less than two seconds. 

To enhance scalability, ICP divides canisters into smaller groups, each running within a subnet with a diverse set of replicas. This approach allows for additional subnets as needed. 

> To learn more about the consensus mechanism of ICP, refer to the [Consensus resource](https://internetcomputer.org/how-it-works/consensus). 

Reviewer: ```suggestion
To learn more about the consensus mechanism of ICP, refer to the [consensus resource](https://internetcomputer.org/how-it-works/consensus). 
```
